### PR TITLE
[SPARK-33739] [SQL] Jobs committed through the S3A Magic committer don't track bytes

### PR DESCRIPTION
**Pull request by steveloughran**
BasicWriteStatsTracker to probe for a custom Xattr if the size of
the generated file is 0 bytes; if found and parseable use that as
the declared length of the output.

The matching Hadoop patch in HADOOP-17414:

* Returns all S3 object headers as XAttr attributes prefixed "header."
* Sets the custom header x-hadoop-s3a-magic-data-length to the length of
  the data in the marker file.

As a result, spark job tracking will correctly report the amount of data uploaded
and yet to materialize.

### Why are the changes needed?

Now that S3 is consistent, it's a lot easier to use the S3A "magic" committer
which redirects a file written to `dest/__magic/job_0011/task_1245/__base/year=2020/output.avro`
to its final destination `dest/year=2020/output.avro` , adding a zero byte marker file at
the end and a json file `dest/__magic/job_0011/task_1245/__base/year=2020/output.avro.pending`
containing all the information for the job committer to complete the upload.

But: the write tracker statictics don't show progress as they measure the length of the
created file, find the marker file and report 0 bytes.
By probing for a specific HTTP header in the marker file and parsing that if
retrieved, the real progress can be reported.

There's a matching change in Hadoop [https://github.com/apache/hadoop/pull/2530](https://github.com/apache/hadoop/pull/2530)
which adds getXAttr API support to the S3A connector and returns the headers; the magic
committer adds the relevant attributes.

If the FS being probed doesn't support the XAttr API, the header is missing
or the value not a positive long then the size of 0 is returned.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

 
New tests in BasicWriteTaskStatsTrackerSuite which use a filter FS to
implement getXAttr on top of LocalFS; this is used to explore the set of
options:
* no XAttr API implementation (existing tests; what callers would see with
  most filesystems)
* no attribute found (HDFS, ABFS without the attribute)
* invalid data of different forms

All of these return Some(0) as file length.

The Hadoop PR verifies XAttr implementation in S3A and that
the commit protocol attaches the header to the files.

External downstream testing has done the full hadoop+spark end
to end operation, with manual review of logs to verify that the
data was successfully collected from the attribute.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:225]: `HADOOP-17414` seems to be not committed yet and is still targeting 3.3.1.
Shall we hold one this PR until we have an official Hadoop release?

Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:225]: I'm developing them in sync so they can be done end to end. Yes, it makes sense for the Hadoop one to be committed,  first -but if reviewing the spark one finds changes which need to be done there (is the header a good name?) then an initial review of this PR will make sure that they are lined up

**Comment by dongjoon-hyun** [comment]
Thank you for updating. I'm looking forward to using Hadoop 3.3.1!

**Comment by steveloughran** [comment]
ok, Hadoop-side PR is in trunk; verifying backport compiles and new tests work then it'll be in branch-3.3. 

This patch is ready for final review/merge

**Review by HyukjinKwon — COMMENTED** [review]
Looks making sense to me

**Comment by HyukjinKwon** [comment]
cc @tgravescs @mridulm FYI

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala
Comment (tgravescs) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:107]: do we want this info or just debug?  seems like I would only care if stats didn't come out but maybe its more useful...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala
Comment (tgravescs) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:119]: do we want this warn instead or perhaps just extend saying file stats may be wrong?

**Review by tgravescs — COMMENTED** [review]
overall looks fine to me, do we need to document certain version of Hadoop this works with and that they will be wrong with?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala
Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:107]: can put to debug.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala
Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:119]: OK. I'd say at info as warn seems overkill for progress report issues.

**Comment by steveloughran** [comment]
I'll update the docs. Also going to cut back on the warnings now that AWS S3 is consistent (And all third party ones were consistent out the box too). This is a good thing for apps like spark (much easier to do workflows across applications), but you still can't safely use the rename committer there as dir rename is non-atomic.
Fortunately, commit-by-rename is so slow people have a reason to switch to a better one, either the S3A ones or that from EMR :)

File: docs/cloud-integration.md
Comment (tgravescs) [inline docs/cloud-integration.md:265]: extra "Are" here - remove

File: docs/cloud-integration.md
Comment (tgravescs) [inline docs/cloud-integration.md:270]: extra Are

Comment (steveloughran) [reply docs/cloud-integration.md:270]: sorry; my fault for having dictation turned on and not paying attention to what the music in the background was saying...

**Review by tgravescs — APPROVED** [review]

File: docs/cloud-integration.md
Comment (dongjoon-hyun) [inline docs/cloud-integration.md:52]: Shall we remove line 60 together?
> 1. The output of work may not be immediately visible to a follow-on query.

Comment (steveloughran) [reply docs/cloud-integration.md:52]: I worry about openstack swift. The original version was woefully inconsistent. Not been near it long enough to know what's change. with swift I could consistently replicate an inconsistency in a few lines

1. write a file of length, say, 1KB
2. overwrite with a shorter dataset, e.g. 512 bytes
3. open file
4. read from 0-128: get the new data
5. read from 768-1024: get the old data

No S3 implementation that I know of (i.e. the open source or commercial ones) are inconsistent, nor have they ever had the 404 caching issue. But people should still be aware of the risk.

oh, and I haven' t played with any of the chinese cloud stores for which connectors now exist. So I can't make statements there. All I can say is the "big three outside china" are consistent.

Comment (HeartSaVioR) [reply docs/cloud-integration.md:52]: The line deletion is based on the fact S3 is now strong consistency, but that also means we only consider these three vendors. Why not explicitly saying it and update the description tied to these vendors? We would never be able to consider all possible implementations and for some minority the description may be wrong. Let's just make it clear.

Comment (HeartSaVioR) [reply docs/cloud-integration.md:52]: Or, what about leaving this line as it is (so that the description enumerates all "possible" issues on object store considering beyond the big three), and elaborate which key differences are no longer in effect with strong consistency in consistency section?

Comment (steveloughran) [reply docs/cloud-integration.md:52]: > (Probably you're considering more like S3 compatible implementations, but you've also mentioned you don't consider the chinese cloud stores so it can't be exhaustive.)

I don't test them; I don't know their consistency

Comment (steveloughran) [reply docs/cloud-integration.md:52]: reviewed the docs, huawei cos and tencent obs both seem consistent. swift is now the outlier. Moving the issue into a paragraph in the "consistency" section

**Comment by steveloughran** [comment]
I've looked at the stores for which hadoop-trunk has connectors, looks like only openstack swift is inconsistent. Moved the details on consistency down, called out swift and said "consult the docs". After all, alternative implementations of the swift API (I'm thinking IBM's work) probably is consistent.

File: docs/cloud-integration.md
Comment (HeartSaVioR) [reply docs/cloud-integration.md:52]: @steveloughran Thanks for the update!
@dongjoon-hyun Could you please check whether the comments and new change make sense to you? Thanks in advance!

**Comment by HeartSaVioR** [comment]
The doc change looks good to me, but I don't feel qualified to review the code. As @tgravescs already reviewed and approved, it sounds more natural if he can make "sign-off".

@tgravescs Kindly reminder. Thanks!

File: docs/cloud-integration.md
Comment (dongjoon-hyun) [inline docs/cloud-integration.md:196]: We can remove this repetition (line 194 ~ 196) because we already have this at the beginning of this section.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BasicWriteStatsTracker.scala:127]: `Xattr` -> `XAttr`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteTaskStatsTrackerSuite.scala
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/BasicWriteTaskStatsTrackerSuite.scala:228]: `Xattr` -> `XAttr`?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM, too.

File: docs/cloud-integration.md
Comment (steveloughran) [reply docs/cloud-integration.md:196]: done

**Comment by steveloughran** [comment]
@tgravescs can you hit the merge button before some other change breaks this PR? thanks

**Comment by tgravescs** [comment]
merged to master, thanks @steveloughran @dongjoon-hyun @HeartSaVioR

**Comment by steveloughran** [comment]
thanks!

**Comment by HyukjinKwon** [comment]
I was here too! 😃

**Pull request closed without merging**
